url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/fsa/tffst/Tffst.java

```diff
@@ -0,0 +1,1712 @@
+   * this o theOther (i.e. theOther(this) )   
+   * 
+   * Constraint: this cannot have e-input labels and theOther cannot have e-output labels.  
+   * 
+   * ct((p1,π1,π1,q1,1),(p2,π2,π2,q2,1))=((p1,p2),π1∧π2,π1∧π2,(q1,q2),1)
+   * ct((p1,φ,π1,q1,0),(p2,π2,π2,q2,1))=((p1,p2),φ,π1∧π2,(q1,q2),0)
+   * ct((p1,π1,π1,q1,1),(p2,π2,ψ,q2,0))=((p1,p2),π1∧π2,ψ,(q1,q2),0)
+   * ct((p1,φ,π1,q1,0),(p2,π2,ψ,q2,0))=((p1,p2),φ,ψ,(q1,q2),0) if
+   * satisfiable(π1∧π2)
+   * 
+   * E= {ct(e1,e2 )|e1 ∈ E1 , e2 ∈ E2 } ∪ {((p1,p2 ), ,ψ,(p1,q2),0)|p1 ∈
+   * Q1,(p2,,ψ,q2,0) ∈ E2 } ∪ {((p1,p2 ),φ, ,(q1,p2),0)|p2 ∈ Q2,(p1,φ,,q1,0) ∈
+   * E1 }
+   * 
+   * 
+   * 
+   * 
+   */
+  
+  public Tffst composition(Tffst theOther) { 
+
+    Tffst thisSimple = this.toSingleLabelTransitions();
+    Tffst theOtherSimple = theOther.toSingleLabelTransitions();
+    
+    Tffst out = new Tffst();
+
+    Set<State> thisStates = thisSimple.getLiveStates();
+    Set<State> theOtherStates = theOtherSimple.getLiveStates();
+
+    HashMap<StatePair, State> newstates = new HashMap<StatePair, State>();
+
+    for (State thisState : thisStates){
+      for (State theOtherState : theOtherStates){
+        
+        // get the new state that is equivalent to this pair, create it. 
+        State outStartState = newstates.get(new StatePair(thisState, theOtherState));
+        if (outStartState == null) {
+          outStartState = new State();
+          newstates.put(new StatePair(thisState, theOtherState), outStartState);
+          if (thisState.equals(thisSimple.initial) && theOtherState.equals(theOtherSimple.initial))
+            out.initial = outStartState;
+          outStartState.setAccept(thisState.isAccept() && theOtherState.isAccept());
+        }
+
+        Set<Transition> thisStateTrans = thisState.getTransitions();
+        Set<Transition> theOtherStateTrans = theOtherState.getTransitions(); 
+        //add fake e/e transition loop  
+        thisStateTrans.add(new Transition(SimpleTf.Epsilon(),SimpleTf.Epsilon(),thisState));
+        theOtherStateTrans.add(new Transition(SimpleTf.Epsilon(),SimpleTf.Epsilon(),theOtherState));
+
+		for (Transition t1 : thisStateTrans) {
+		  for (Transition t2 : theOtherStateTrans) {
+			boolean go = false; 
+
+			if ((t1.labelOut.isEpsilon() && t2.labelIn.isEpsilon())) go = true;
+			if ((!t1.labelOut.isEpsilon() && !t2.labelIn.isEpsilon())) 
+			  if (t1.labelOut.get(0).andSimple(t2.labelIn.get(0)).satisfiable()) go = true; 
+			  
+			if (go) { 
+			  // get the new state that is equivalent to this pair, create it.
+			  State outToState = newstates.get(new StatePair(t1.getTo(), t2.getTo()));
+			  if (outToState == null) {
+				outToState = new State();
+				newstates.put(new StatePair(t1.getTo(), t2.getTo()), outToState);
+				if (t1.getTo().equals(thisSimple.initial) && t2.getTo().equals(theOtherSimple.initial))
+				  out.initial = outToState;
+				outToState.setAccept(t1.getTo().isAccept() && t2.getTo().isAccept());
+			  }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 376:382, 526:532.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.